src/Framework/TestSuite.php

```diff
@@ -797,55 +796,59 @@ public function warnings(): array
     }
 
     /**
-     * @return array<callable-string>
+     * @return list<ExecutionOrderDependency>
      */
     public function provides(): array
     {
         if ($this->providedTests === null) {
             $this->providedTests = [];
 
-            $callableName = $this->getName() . '::class';
-
-            if (is_callable($callableName, true)) {
-                $this->providedTests = [$callableName];
+            if (is_callable($this->sortId(), true)) {
+                $this->providedTests[] = new ExecutionOrderDependency($this->sortId());
             }
 
             foreach ($this->tests as $test) {
-                if (!($test instanceof self || $test instanceof TestCase)) {
+                if (!($test instanceof Reorderable)) {
                     // @codeCoverageIgnoreStart
                     continue;
                     // @codeCoverageIgnoreEnd
                 }
-                $this->providedTests = array_merge($this->providedTests, $test->provides());
+                $this->providedTests = ExecutionOrderDependency::mergeUnique($this->providedTests, $test->provides());
             }
         }
 
         return $this->providedTests;
     }
 
     /**
-     * @return array<callable-string>
+     * @return list<ExecutionOrderDependency>
      */
     public function requires(): array
     {
         if ($this->requiredTests === null) {
             $this->requiredTests = [];
 
             foreach ($this->tests as $test) {
-                if (!($test instanceof self || $test instanceof TestCase)) {
+                if (!($test instanceof Reorderable)) {
                     // @codeCoverageIgnoreStart
                     continue;
                     // @codeCoverageIgnoreEnd
                 }
-                $this->requiredTests = array_merge($this->requiredTests, $test->requires());
+                $this->requiredTests = ExecutionOrderDependency::filterInvalid($this->requiredTests);
+                $this->requiredTests = ExecutionOrderDependency::mergeUnique($this->requiredTests, $test->requires());
```

Comment:
Instead of reassigning twice, what do you think about this?

```suggestion
                $this->requiredTests = ExecutionOrderDependency::mergeUnique(
                    ExecutionOrderDependency::filterInvalid($this->requiredTests),
                    $test->requires()
                );
```

Reply:
Nicer to read, I agree.